src/ring_theory/tensor_product.lean

```diff
@@ -1,36 +1,321 @@
+    simp only [algebra_tensor_module.smul_def, alg_hom.map_smul, smul_apply, rtensor_smul]
+  end }
+
+end semiring
+
+section comm_semiring
+variables [comm_semiring R] [comm_semiring A] [algebra R A]
+variables [add_comm_monoid M] [semimodule R M] [semimodule A M] [is_scalar_tower R A M]
+variables [add_comm_monoid N] [semimodule R N]
+variables [add_comm_monoid P] [semimodule R P] [semimodule A P] [is_scalar_tower R A P]
+
+variables {R A M N P}
+/-- Heterobasic version of `tensor_product.lift`:
+
+Constructing a linear map `M ⊗[R] N →[A] P` given a bilinear map `M →[A] N →[R] P` with the
+property that its composition with the canonical bilinear map `M →[A] N →[R] M ⊗[R] N` is
+the given bilinear map `M →[A] N →[R] P`. -/
+@[simps] def lift' (f : M →ₗ[A] (N →ₗ[R] P)) : (M ⊗[R] N) →ₗ[A] P :=
+{ map_smul' := λ c, show ∀ x : M ⊗[R] N, (lift (f.restrict_scalars R)).comp (lsmul R _ c) x =
+      (lsmul R _ c).comp (lift (f.restrict_scalars R)) x,
+    from ext_iff.1 $ tensor_product.ext $ λ x y,
+    by simp only [comp_apply, algebra.lsmul_coe, smul_tmul, lift.tmul, coe_restrict_scalars_eq_coe,
+        f.map_smul, smul_apply'],
+  .. lift (f.restrict_scalars R) }
+
+@[simp] lemma lift'_tmul (f : M →ₗ[A] (N →ₗ[R] P)) (x : M) (y : N) :
+  lift' f (x ⊗ₜ y) = f x y :=
+lift.tmul' x y
+
+/-- Heterobasic version of `tensor_product.curry`:
+
+Given a linear map `M ⊗[R] N →[A] P`, compose it with the canonical
+bilinear map `M →[A] N →[R] M ⊗[R] N` to form a bilinear map `M →[A] N →[R] P`. -/
+@[simps] def curry' (f : (M ⊗[R] N) →ₗ[A] P) : M →ₗ[A] (N →ₗ[R] P) :=
+{ map_smul' := λ c x, linear_map.ext $ λ y, f.map_smul c (x ⊗ₜ y),
+  .. curry (f.restrict_scalars R) }
+
+variables (R A M N P)
+/-- Heterobasic version of `tensor_product.uncurry`:
+
+Linearly constructing a linear map `M ⊗[R] N →[A] P` given a bilinear map `M →[A] N →[R] P`
+with the property that its composition with the canonical bilinear map `M →[A] N →[R] M ⊗[R] N` is
+the given bilinear map `M →[A] N →[R] P`. -/
+@[simps] def uncurry' : (M →ₗ[A] (N →ₗ[R] P)) →ₗ[A] ((M ⊗[R] N) →ₗ[A] P) :=
+{ to_fun := lift',
+  map_add' := λ f g, ext $ λ x y, by simp only [lift'_tmul, add_apply],
+  map_smul' := λ c f, ext $ λ x y, by simp only [lift'_tmul, smul_apply'] }
+
+/-- Heterobasic version of `tensor_product.lcurry`:
+
+Given a linear map `M ⊗[R] N →[A] P`, compose it with the canonical
+bilinear map `M →[A] N →[R] M ⊗[R] N` to form a bilinear map `M →[A] N →[R] P`. -/
+@[simps] def lcurry' : ((M ⊗[R] N) →ₗ[A] P) →ₗ[A] (M →ₗ[A] (N →ₗ[R] P)) :=
+{ to_fun := curry',
+  map_add' := λ f g, rfl,
+  map_smul' := λ c f, rfl }
+
+/-- Heterobasic version of `tensor_product.lift_equiv`:
+
+A linear equivalence constructing a linear map `M ⊗[R] N →[A] P` given a
+bilinear map `M →[A] N →[R] P` with the property that its composition with the
+canonical bilinear map `M →[A] N →[R] M ⊗[R] N` is the given bilinear map `M →[A] N →[R] P`. -/
+def lift_equiv' : (M →ₗ[A] (N →ₗ[R] P)) ≃ₗ[A] ((M ⊗[R] N) →ₗ[A] P) :=
```

Comment:
I think the other version is with a dot not an underscore?
```suggestion
/-- Heterobasic version of `tensor_product.lift.equiv`:

A linear equivalence constructing a linear map `M ⊗[R] N →[A] P` given a
bilinear map `M →[A] N →[R] P` with the property that its composition with the
canonical bilinear map `M →[A] N →[R] M ⊗[R] N` is the given bilinear map `M →[A] N →[R] P`. -/
def lift.equiv' : (M →ₗ[A] (N →ₗ[R] P)) ≃ₗ[A] ((M ⊗[R] N) →ₗ[A] P) :=
```

Reply:
Done.